### ISSUE TITLE
Tracking progress of Datasets & Entities

### ISSUE THREAD

**Issue by ktuite** [getodk/central-backend#649]
This issue tracks progress of the initial implementation of Datasets and Entities, including PRs, to-dos, and issues/bugs that we discover along the way. It is patterned off of a frontend [issue](https://github.com/getodk/central-frontend/issues/565) about a large code migration.

## `entities` branch

PR #576 represents the `entities` branch, a large initial chunk of work that covers:
* Parsing XML
	* Parsing Form XML to get dataset information
		* Validation
	* Parsing Submission XML to make entities
		* Validation
* Dataset management
	* New schema:
		* Dataset
		* Dataset Properties
		* Dataset Property Fields
	* Create/update datasets on form uploads
	* New API endpoints:
		* List datasets
		* View dataset diff (what effect will publishing a draft form have on a dataset)
		* Link/unlink a dataset to a from that consumes a dataset as an attachment
* Entity management
	* New schema
		* Entity
		* Entity def
	* Worker that looks for approved submissions and tries to make entities out of them
* Entity/dataset CSV serving
	* Stream out entities as CSV for forms to use as attachments
* Audit logging
	* Logging dataset create/update events
	* Logging entity create events
	* Exposing dataset and entity actions through API 

### Things to get into this PR before merging
* [x] Merge #648 
* [x] Merge #646 
* [x] Revisit `exists` vs `blobExists` attachment API ([comments](https://github.com/getodk/central-backend/pull/576#discussion_r1008407467))
* [x] Merge migrations into one feature migration
* [x] Delete some unnecessary tests that were intermediaries before API existed
* [x] Move other test around 
* [x] Merge #653 

<hr>

Critical to do before release:
- [x] https://github.com/getodk/central-backend/milestone/3
- [x] Update dataset/entity parsing to reflect latest version of spec. This also includes
  * dataset name validation: either trim whitespace or don't allow names with whitespace (right now it lets them through untrimmed)
  * go back to making dataset names case sensitive 
- [x] Make sure form deletion and form purging cooperate with entities 
- [x] Check handling of encrypted forms
- [x] Docs: #656 

Likely before release:
- [x] Revisit entity-insertion pre-check query ([here](https://github.com/getodk/central-backend/pull/576#discussion_r1008419949))
- [x] Figure out new names for Datasets query module functions `getByName` and `getByProjectAndName` ([here](https://github.com/getodk/central-backend/pull/576#discussion_r1008419220))
- [x] Revisit Actee provisioning on Dataset creation via SQL ([here](https://github.com/getodk/central-backend/pull/576#discussion_r1008414409))

Collecting issues / unexpected behavior / bugs:
* [x] Making an entity out of a submission that is from a form that isn't actively linked to a dataset (saw this when rolling back and reapplying a migration, may be hard to recreate this bug but could point so some other underlying issue)
* [x] Bug: re-approving a submission and re-making the entity makes things crash :( 

<hr>

To come back to after release:
* Entity deletion
* How deleting forms and submissions impacts datasets and entities 
* More efficient streaming of entities
* #660
* #661
* #664
* getodk/central#348

**Comment by matthew-white** [comment]
Also for the API docs: adding the new audit log actions to the list of actions.

**Comment by matthew-white** [comment]
Maybe this is done already, but we also talked about verifying that using encryption + entities together won't cause Backend to crash (even if it's not an amazing user experience):

- It should be possible to upload a form that both uses encryption and creates entities: there should be no error.
- If a submission is created for such a form, that shouldn't result in an entity, since Backend can't read the contents of the submission. However, Backend should be able to handle the submission reasonably well and not crash.

There are different ways that a form can use encryption. If a project as a whole uses managed encryption, then forms within the project will be encrypted. A form can also specify its own encryption even if it's within a project that doesn't use managed encryption.

**Comment by matthew-white** [comment]
Should we validate the type of fields that can be used as entity properties? My understanding is that groups and repeat groups are fields that can't be used as entity properties.

Less sure about this one: for now, should we prohibit any fields within a repeat group from being used as entity properties?

**Comment by matthew-white** [comment]
It looks like it's currently not possible to purge forms if a soft-deleted form has a submission that has created an entity. I wrote a test along those lines here:

```js
it('should allow a form that has created an entity to be purged', testService((service, container) =>
  service.login('alice', (asAlice) =>
    asAlice.post('/v1/projects/1/forms?publish=true')
      .set('Content-Type', 'application/xml')
      .send(testData.forms.simpleEntity)
      .expect(200)
      .then(() => asAlice.post('/v1/projects/1/forms/simpleEntity/submissions')
        .send(testData.instances.simpleEntity.one)
        .set('Content-Type', 'application/xml')
        .expect(200))
      .then(() => asAlice.patch('/v1/projects/1/forms/simpleEntity/submissions/one')
        .send({ reviewState: 'approved' })
        .expect(200))
      .then(() => exhaust(container))
      .then(() => asAlice.delete('/v1/projects/1/forms/simpleEntity')
        .expect(200))
      .then(() => container.Forms.purge(true)))));
```

Right now the test fails with:

error: update or delete on table "submission_defs" violates foreign key constraint "entity_defs_submissiondefid_foreign" on table "entity_defs"

**Comment by matthew-white** [comment]
Here's another one related to form purging, more of an edge case. If all the forms related to a dataset are purged, then the dataset is no longer listed, even though it might still be used in active forms as a form attachment. When we list datasets, we filter out datasets from form drafts that haven't been published yet. But I think the result of that filtering is also this behavior around purging.

Here's the relevant SQL:

https://github.com/getodk/central-backend/blob/230dbb5511d7c325319109c0c7fb70ba538ccb20/lib/model/query/datasets.js#L55-L60

Here's a try at a test along these lines:

```js
it('should list a dataset even if related forms have been purged', testService((service, container) =>
  service.login('alice', (asAlice) =>
    asAlice.post('/v1/projects/1/forms?publish=true')
      .set('Content-Type', 'application/xml')
      .send(testData.forms.simpleEntity)
      .expect(200)
      .then(() => asAlice.get('/v1/projects/1/datasets')
        .expect(200)
        .then(({ body }) => {
          body.length.should.equal(1);
        }))
      .then(() => asAlice.delete('/v1/projects/1/forms/simpleEntity')
        .expect(200))
      .then(() => container.Forms.purge(true))
      .then(() => asAlice.get('/v1/projects/1/datasets')
        .expect(200)
        .then(({ body }) => {
          body.length.should.equal(1);
        })))));
```

To fix this, I think we may need to add a "published" flag to the `datasets` table. That would allow us to recognize a dataset as having been associated with a published form even if that published form is purged.

**Comment by matthew-white** [comment]
I've been taking a look at #576 and noticed these two remaining related TODOs from #633:

https://github.com/getodk/central-backend/blob/230dbb5511d7c325319109c0c7fb70ba538ccb20/lib/resources/datasets.js#L27

https://github.com/getodk/central-backend/blob/230dbb5511d7c325319109c0c7fb70ba538ccb20/test/integration/api/datasets.js#L85-L93

**Comment by matthew-white** [comment]
If the form specifies a dataset name with leading or trailing white space, then `validateDatasetName()` will trim the name before validating it. However, `getDataset()` doesn't seem to trim it when it returns it, so the dataset name ends up having white space. I'd expect that either `validateDatasetName()` would reject such a name or that `getDataset()` would also trim it. I wrote a test along those lines:

```js
it('should trim the dataset name', testService((service) =>
  service.login('alice', (asAlice) =>
    asAlice.post('/v1/projects/1/forms?publish=true')
      .set('Content-Type', 'application/xml')
      .send(testData.forms.simpleEntity.replace('people', ' people\n'))
      .expect(200)
      .then(() => asAlice.get('/v1/projects/1/datasets')
        .expect(200)
        .then(({ body }) => {
          body.length.should.equal(1);
          body[0].name.should.equal('people');
        })))));
```

**Comment by matthew-white** [comment]
Another one about dataset name. I'm thinking that we shouldn't make that case-insensitive. xmlFormId and form attachment name are case-sensitive, so I think it makes sense for dataset name to be so as well. Dataset names can contain any Unicode letter, but I think it's difficult to combine that with case-insensitivity. For example, `toLowerCase()` doesn't return the correct result in Turkish for İ (dotted I).

**Comment by matthew-white** [comment]
I don't remember that much about itemsets, but given that we have special logic around form attachments named `itemsets.csv`, I think we should forbid datasets from being named `itemsets`.

**Comment by issa-tseng** [comment]
the specialness is just that itemsets are not declared any normal way in the xml. they are ultimately used the same way and i think it’s probably a good feature for people to be able to make a dataset their itemset. but ln would know for sure.

**Comment by lognaturel** [comment]
It feels to me like `itemsets` would be a really strange dataset name so I doubt there would be collisions. `itemsets` is plural because it essentially lets you define multiple datasets that are processed in a custom/magic way by Collect. As @issa-tseng says, there's nothing special about the `itemsets.csv` filename other than IF you're using "fast external itemsets" then that's the name of the CSV that's produced. I don't think it needs to be treated in any special way but it's a good thing to bring up!

**Comment by matthew-white** [comment]
I think I'm seeing that simply creating a form draft has the effect of increasing the `revisionNumber` of a dataset. For example:

- Create a new form that creates entities for a new dataset named `people`. Publish the form. &hellip;/datasets returns `people` with a `revisionNumber` of 0.
- Create a new draft for the form, but don't publish it. Now &hellip;/datasets returns `people` with a `revisionNumber` of 1.
- Abandon the draft, then create a new draft. &hellip;/datasets returns `people` with a `revisionNumber` of 2.

I'd expect that creating but not publishing a form draft wouldn't change an existing dataset in a way that's visible over the API.

**Comment by matthew-white** [comment]
Relatedly, we discussed while reviewing #576 interactively that the timing of the `dataset.create` and `dataset.update` audit log entries might be surprising:

- If you upload but don't publish a form definition that would create entities for a new dataset, then `dataset.create` is immediately logged, even if the form draft is ultimately abandoned and the dataset never becomes visible on the project.
- When I follow the steps in my previous comment, whenever the `revisionNumber` increases, there is also a `dataset.update` audit log entry.

I'd expect `dataset.create` and `dataset.update` to be logged only when the associated form draft is published. I know that creating a draft can have an immediate effect on the `datasets` table, and I think it makes sense that it works that way under the hood. But I think we're exposing those mechanics over the audit log in a way that could be surprising and that ultimately makes the audit log less actionable. It also results in a mismatch between what's in the audit log and what's returned by &hellip;/datasets and &hellip;/dataset-diff.

How about we don't set `.audit` on `Datasets.createOrMerge`, but instead wait until publishing the form to log `dataset.create` or `dataset.update`? I also think it'd be nice to log `dataset.update` only when the dataset has actually changed, not whenever a new form version is published. Could we use the dataset diff to help make these changes?

**Comment by matthew-white** [comment]
I'm noticing the following:

- Upload but don't publish a form definition that would create entities for a new dataset named `people`.
- &hellip;/datasets does not return `people`.
- However, &hellip;/datasets/people/download does serve the entities CSV file. I'd expect that endpoint to return a 404 until the form definition is published.

**Comment by matthew-white** [comment]
There seems to be some inconsistency around how we handle "draft Datasets" and form attachments:

- Upload but don't publish a form definition that would create entities for a new dataset named `people`.
- Upload a form definition for a different form that uses a form attachment named `people.csv`.
- `people.csv` is not autolinked to the `people` dataset, which is what I would expect.
- However, it looks like you can use the PATCH endpoint to link `people.csv` to the unpublished `people` dataset. I'd expect that endpoint to return a 404 until the first form definition is published.

**Comment by matthew-white** [comment]
More of a question: I know that we validate dataset names, but do we also validate property names? I was able to create a property named `÷`, which we wouldn't allow as a dataset name. Is that expected?

**Comment by matthew-white** [comment]
Another one about form drafts + datasets. I'm seeing that as soon as you upload a form definition that would add a property to an existing dataset, that property is included in the entities CSV file. Even if you abandon the form draft, the property is still included in the CSV file. I'd expect that a property is added to the CSV file only once the draft is published.

**Comment by lognaturel** [comment]
One more thing to add under " Update dataset/entity parsing to reflect latest version of spec": we need to reject form definitions that are written for future spec versions. More [here](https://docs.google.com/document/d/1-O_wT0JZZFaWvzHl_krrzERJcnrdGMKP55ZZ_hxV0gU/edit#heading=h.pilfnyfgdjnn).

We're going to release the first spec as 2022.1.0. The patch version is there to keep track of changes that don't affect compatibility (e.g. wording changes to the spec document, optional additions). So anything that doesn't start with 2022.1 should be rejected.

**Comment by matthew-white** [comment]
Do we want an `order` column on `ds_properties`? I might be missing something, but right now when we write the entities CSV file, I don't see that there's a guarantee about the order of the columns after `name` and `label` (the order of the properties). I'd expect that the order of the properties would be stable.

I'm not sure how much the column order matters when the entities CSV file is served to an OpenRosa client. The column order isn't user-facing there, right? But as users start downloading the entities CSV file for use in data management or analysis, I think it'll be important that the column order be stable.

I don't think we can rely on `form_fields.order` for this, as multiple forms can feed into the dataset: two forms may save the same fields to the same dataset properties, yet have those fields in different relative orders. Even for a single form, the field order may change from version to version. Also, given that forms and datasets are allowed to have separate lifecycles, I think that the order of properties should remain stable even if a contributing form is deleted.

Instead, I think we should independently track the property order using a column on `ds_properties`. Maybe for now we just order by `ds_properties.id` and have that be our guaranteed order.

**Comment by ktuite** [comment]
Closing this issue because everything mentioned in here is either completed 🎉 or captured in a new issue.

**Comment by lognaturel** [comment]
🎉 🎉 👏 👏 👏 🎉 🎉
